sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/ServiceBusReceivedMessageExtensions.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Azure.Messaging.ServiceBus.Amqp
+{
+    /// <summary>
+    /// Extension methods for <see cref="ServiceBusReceivedMessage"/>.
+    /// </summary>
+    public static class ServiceBusReceivedMessageExtensions
+    {
+        /// <summary>
+        /// Gets the body type from an AMQP <see cref="ServiceBusReceivedMessage"/>.
+        /// </summary>
+        /// <param name="message">The message to use.</param>
+        /// <returns>Returns the body type of the AMQP message.</returns>
+        public static AmqpBodyType GetAmqpBodyType(this ServiceBusReceivedMessage message)
+        {
+            var transportBody = (AmqpTransportBody)message.SentMessage.TransportBody;
+            // If DataCount is greater than 0, it means that data was received, and the message is non-Sequence/Value.
+            // This won't be greater than 0 on user-constructed Data messages, but it will be on received Data messages.
+            // The actual AmqpBodyType will be Unspecified on received Data messages with 0 or 1 data elements.
+            // Therefore, a received message can be converted to a sent message, and the AmqpBodyType will be correct for that process.
+            // However, we let the user know the actual AmqpBodyType here as they would expect.
+            return transportBody.DataCount > 0 ? AmqpBodyType.Data : message.SentMessage.GetAmqpBodyType();
```

Comment:
Why can't we just directly return `message.SentMessage.GetAmqpBodyType()` without the DataCount check?
Won't this be populated for Data messages?

Reply:
The body type will be misaligned because of implicit understanding of body type on received messages. Meaning, I sent a message via Data, and I receive that message and it won't be Data. So, this check makes that alignment correct. The internal body type will be Unspecified for the `SentMessage` so that you can convert a received message to a sent message properly. However, the user wants to see Data here since that's what it is for their purposes. If we actually changed the body type to Data for the 1 data element situation, Data itself is null, because only the Body has the information. We consider Data on receive to be 2 or more data elements. So, conversion from ServiceBusReceivedMessage to ServiceBusMessage would fail. This seems like the best place to do this magic, since only the user cares that it should be Data instead of Unspecified.

Reply:
This logic has been removed and simplified. DataCount is no longer needed to make judgements about Data.